RFCs/FS-1060-nullable-reference-types.md

```diff
@@ -0,0 +1,823 @@
+
+A warning is given when casting from `C<'S>` to `C<'T | null>` and from `C<'S | null>` to `C<'T>`.
+
+### Checking of non-null references
+
+#### Null assignment and passing
+
+A warning is given if `null` is assigned to a non-null value or passed as a parameter where a non-null reference type is expected:
+
+```fsharp
+let mutable s: string = "hello"
+s <- null // WARNING
+
+let s2: string = null // WARNING
+
+let f (s: string) = ()
+f null // WARNING
+```
+
+#### F# Collection initialization
+
+A warning is given if `null` is used in an F# array/list/seq expression, where the type is already annotated to be non-nullable:
+
+```fsharp
+let xs: string[] = [| ""; ""; null |] // WARNING
+let ys: string list = [ ""; ""; null ] // WARNING
+let zs: seq<string> = seq { yield ""; yield ""; yield null } // WARNING
+```
+
+#### Constructor initialization
+
+Today, it is already a compile error if all fields in an F# class are not initialized by constructor calls. This behavior is unchanged. See [Compatibility with Microsoft.FSharp.Core.DefaultValueAttribute](FS-1060-nullable-reference-types.md#compatibility-with-microsoft.fsharp.core.defaultValueAttribute) for considerations about code that uses `[<DefaultValue>]` with reference types today, which produces `null`.
+
+### Generics
+
+A type parameter is assumed to have non-nullable constraints when declared "normally":
+
+```fsharp
+type C<'T>() = // 'T is non-nullable
+class end
+```
+
+A warning is given if a nullable reference type is used to parameterize another type `C` that  does not accept nullable reference types:
+
+```fsharp
+type C<'T>() = class end
+
+let c = C<string | null> // WARNING
+```
+
+#### Constraints
+
+Today, there are two relevant constraints in F# - `null` and `not struct`:
+
+```fsharp
+type C1<'T when 'T: null>() = class end
+type C2<'T when 'T: not struct>() = class end
+```
+
+Both are actually the same thing for interoperation purposes, as they emit `class` as a constraint. See [Unresolved questions](nullable-reference-types.md#unresolved-questions) for more.
+
+If an F# class `B` has nullability as a constraint, then an inheriting class `D` also inherits this constraint, as per existing inheritance rules:
+
+```fsharp
+type B<'T | null>() = class end
+type D<'T | null>()
+    inherit B<'T>
+```
+
+That is, nullabulity constraints propagate via inheritance.
+
+If an unconstrained type parameter comes from an older C# assembly, it is assumed to be nullable, and warnings are given when non-nullability is assumed. If it comes from F# and it does not have the `null` constraint, then it is assumed to be non-nullable. If it does, then it is assumed to be nullable.
+
+Additionally, C# will introduce the `class?` constraint. This syntax sugar will compile into the `class` constraint, but the class itself will be decorated with an attribute that tells consumers that the `'T` is a nullable reference type. This attribute will need to be respected by the F# compiler.
```

Comment:
When you say "the class itself" I'm not sure what you mean.  Do you mean the generic construct (class, method, struct, whatever) gets an attribute? Or the generic type paramater gets at attribute? This should probably go into the .NET metadata section above as we will need an exact spec of what we're interoperating with

Anyway from the F# perspective this feels the same as `type C<'T when 'T : null>`.